feat: add C implementation for `math/base/special/heaviside`

anandkaranubc opened the pull request:
Progresses https://github.com/stdlib-js/stdlib/issues/649

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the complete C implementation for `math/base/special/heaviside`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progresses https://github.com/stdlib-js/stdlib/issues/649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: initial commit for feedback

anandkaranubc committed:
feat: add support for enums

anandkaranubc committed:
docs: fix paramater name

anandkaranubc committed:
feat: remove enum2str

anandkaranubc committed:
docs: fix comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: add tests

anandkaranubc committed:
feat: add C implementation for `math/base/special/heaviside`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc committed:
style: fix lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/heaviside/include/stdlib/math/base/special/heaviside.h:
```diff
@@ -0,0 +1,53 @@
+#endif
+
+// Enumeration of function continuity:
+typedef enum STDLIB_MATH_CONTINUITY {
```
`STDLIB_BASE_HEAVISIDE_CONTINUITY`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/heaviside/include/stdlib/math/base/special/heaviside.h:
```diff
@@ -0,0 +1,53 @@
+// Enumeration of function continuity:
+typedef enum STDLIB_MATH_CONTINUITY {
+	// Half-maximum:
+	STDLIB_MATH_HALF_MAXIMUM = 0,
```
`STDLIB_BASE_HEAVISIDE_CONTINUITY_HALF_MAXIMUM`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/heaviside/src/addon.c:
```diff
@@ -0,0 +1,42 @@
+	STDLIB_NAPI_ARGV( env, info, argv, argc, 2 );
+	STDLIB_NAPI_ARGV_DOUBLE( env, x, argv, 0 );
+	STDLIB_NAPI_ARGV_INT32( env, continuity, argv, 1 );
+	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_base_heaviside( x, (STDLIB_MATH_CONTINUITY)continuity ), out );
```
```suggestion
	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_base_heaviside( x, continuity ), out );
```

anandkaranubc committed:
fix: enum names

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/heaviside

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/heaviside/lib/str2enum.js:
```diff
@@ -0,0 +1,52 @@
+*/
+function str2enum( ctype ) {
+	var v = ENUM[ ctype ];
+	return ( typeof v === 'number' ) ? v : null; // note: we include this guard to prevent walking the prototype chain
```
Potential discrepancy: when supplying an invalid continuity string, we return `null` here, which causes an error to be thrown when invoking the native add-on. For the native add-on to behave like the JS implementation, I propose we pass down `3` instead of `null` here so that any other continuity string will be treated as the discontinuity case.

anandkaranubc replied:
@Planeshifter, right! Do you think it's better to return `-1` here instead of `null` so that it triggers the default block in the switch statement?

Also, for the native add-on, don't we build it in a way that matches the C implementation? For example, in the implementation of `fibonacci`, let's say in [test.native.js](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/test/test.native.js), we removed the test where the input is not an integer. Still, having said that, your point stands. A user can pass in any number in `heaviside` for the continuity option other than 0, 1, 2, or 3 in `C`, and it should return NaN.

anandkaranubc replied:
If the above looks right, I will also update the implementation, tests and docs accordingly... Thanks for the feedback!

kgryte replied:
Returning `-1` instead of `null` is fine.

Planeshifter replied:
`-1` sounds good to me, as well. It's preferable to `3` to not have a magic value in here that could drift from the enum in case we, for whatever reason, reorder the enum values.

anandkaranubc replied:
On it! Thank you for the confirmations!

Planeshifter reviewed (dismissed):
Looks good to me overall; one question.

anandkaranubc committed:
Merge branch 'develop' into feat/heaviside

anandkaranubc committed:
refactor: change fallback return value from null to -1 in str2enum

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
```diff
@@ -214,6 +214,8 @@ The `continuity` parameter may be one of the following values:
 -   `STDLIB_BASE_HEAVISIDE_CONTINUITY_RIGHT_CONTINUOUS`: if `x == 0`, the function returns `1.0`.
 -   `STDLIB_BASE_HEAVISIDE_CONTINUITY_DISCONTINUOUS`: if `x == 0`, the function returns `NaN`.
 
+If not provided a valid `continuity` parameter, the function returns `NaN` for `x == 0`, behaving like the discontinuous case.
```
Does this line look correct then?

kgryte replied:
Did you confirm that C allows the continuity parameter to be a value other than an allowed enum value?

anandkaranubc replied:
@kgryte It behaves like the discontinuous case if you pass a number (other than 0,1,2,3), say, `-1` as the continuity parameter. If you pass something like `FOO`, it throws an error: *use of undeclared identifier 'FOO'*.

kgryte commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
```diff
@@ -164,6 +164,104 @@ logEachMap( 'H(%0.4f) = %0.4f', x, heaviside );
+-   `STDLIB_BASE_HEAVISIDE_CONTINUITY_RIGHT_CONTINUOUS`: if `x == 0`, the function returns `1.0`.
+-   `STDLIB_BASE_HEAVISIDE_CONTINUITY_DISCONTINUOUS`: if `x == 0`, the function returns `NaN`.
+
+If not provided a valid `continuity` parameter, the function returns `NaN` for `x == 0`, behaving like the discontinuous case.
```
```suggestion
If provided a `continuity` argument which is not one of the enumeration constants listed above, the function returns `NaN` for `x == 0`, behaving like the discontinuous case.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/heaviside/README.md:
```diff
@@ -164,6 +164,104 @@ logEachMap( 'H(%0.4f) = %0.4f', x, heaviside );
+The function accepts the following arguments:
+
+-   **x**: `[in] double` input value.
+-   **continuity**: `[in] STDLIB_BASE_HEAVISIDE_CONTINUITY` continuity option
```
```suggestion
-   **continuity**: `[in] STDLIB_BASE_HEAVISIDE_CONTINUITY` continuity option.
```

anandkaranubc committed:
Update lib/node_modules/@stdlib/math/base/special/heaviside/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc committed:
docs: update the README

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc committed:
test: add test for the default case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: update tests to match JS tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/heaviside

anandkaranubc committed:
refactor: remove `typedef` and update Makefile

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/heaviside/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = ( 100.0*rand_double() ) - 50.0;
```
Can we generate this value via a `random_uniform` function, please?

anandkaranubc committed:
Merge branch 'develop' into feat/heaviside

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/heaviside/lib/native.js:
```diff
@@ -0,0 +1,64 @@
+*
+* @private
+* @param {number} x - input value
+* @param {string} [continuity] - continuity option
```
Should we mark the parameter as mandatory here?

```suggestion
* @param {string} continuity - continuity option
```

Planeshifter replied:
The JS implementation does have it as optional, too, so probably better to leave as-is for now. There is an argument to be made to make the parameter mandatory overall in this package as we have moved to avoid variadic functions for performance reasons, but that could be handled in a separate PR.

cc @kgryte

kgryte replied:
I suggest going ahead and making mandatory across the package.

anandkaranubc replied:
Cool! Updating the PR accordingly....

Thanks @Planeshifter and @kgryte for the feedback!

anandkaranubc committed:
bench: use `random_uniform`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
LGTM!

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/heaviside/test/test.native.js:
```diff
@@ -0,0 +1,164 @@
+	t.end();
+});
+
+tape( 'the function returns `1` if `x` is positive', opts, function test( t ) {
+	var x;
+	var v;
+	var i;
+
+	for ( i = 0; i < 1e3; i++ ) {
+		x = ( randu()*100.0 ) + EPS;
+		v = heaviside( x );
+		t.strictEqual( v, 1.0, 'returns expected value' );
+	}
+	t.end();
+});
```
@Planeshifter As far as I remember, I kept the `continuity` parameter as optional because the `str2enum` function handles the condition when it's missing. 

Very open to update this too if we need to make the second parameter as mandatory.

kgryte commented:
Ref: https://github.com/stdlib-js/stdlib/pull/10196
